Fix: Address bar URL is not updated after login or landing on root

kidroca opened the pull request:
<If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit.>
@marcaaron 

### Details
Update navigation initial state logic. 

### Fixed Issues
Fixes #2325

### Tests

All the steps described here are for web and mWeb. For desktop and the native apps there should be no regressions. E.g. opening the mobile app should still go to the Home screen where you select a conversation.

Best to use Incognito Mode for the below 2 flows so that you're sure something in existing storage doesn't help in passing these steps

#### Login Flow
1. Start from logged out state on `/` (with no traces of previous report in the url)
2. Log In
3. Observe the URL updating to the currently loaded report - e.g. `/r/71403817` 

#### Create Account
1. Start from logged out state and proceed to create a new account
2. On dev I'm copying the email link and appending it to localhost - e.g. http://localhost:8080//setpassword/10024706/EHOHMPGWM
3. After successfully creating an account you should see the URL representing the currently viewed report  - e.g. `/r/71403817` 

#### Opening the app when authenticated
1. Be logged in
2. Close the app window/tab
3. Open a browser window and go to the root address of the app (for me it's `localhost:8080`)
4. Observe the URL updating to the currently loaded report - e.g. `/r/71403817` 

#### Navigating to the root of the app
1. Be logged in
2. Edit the address bar to go to the root of the app - e.g. type  `localhost:8080/`
3. Observe the URL updating to the currently loaded report - e.g. `/r/71403817` 

#### Refreshing
1. Be logged in
2. Refresh the page
3. The url should not change, it should still show the currently loaded report - e.g. `/r/71403817` 



### QA Steps
Same as the tests above

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web

https://user-images.githubusercontent.com/12156624/114418150-e568b900-9bba-11eb-9519-089c7086213e.mp4

#### Mobile Web

https://user-images.githubusercontent.com/12156624/114418277-06c9a500-9bbb-11eb-8f87-5397e07b8f39.mp4

kidroca committed:
refactor: NavigationRoot extract `renderFallback`

kidroca committed:
refactor: NavigationRoot extract change handler

kidroca committed:
refactor: NavigationRoot remove Linking handling

With the current configuration it should be possible for the
navigation to resolve initial state entirely through `linkingConfig`

kidroca committed:
feat: AuthScreens setup initial Home screen params

1. Either initial report data is available right from the start from Onyx
2. Or it is fetched and restored by `fetchAllReports`
3. Or if all else fails start with no report opened

kidroca commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -109,11 +118,20 @@ class AuthScreens extends React.Component {
+        // Update initialHomeParams only once
+        if (!this.initialHomeParams) {
+            // Either we have a reportID or fetchAllReports resolved with no reports. Otherwise keep waiting
+            if (nextProps.initialReportID || nextProps.initialReportID === '') {
+                this.setInitialHomeParams(nextProps.initialReportID);
+                return true;
+            }
```
@marcaaron can you confirm this would work

The idea is:
1. If we have a stored report we'll never end up in here. It'll be the initial value.
2. In case we don't we'll end up here after `fetchAllReports` ends
3. Looking at the code `fetchAllReports` would never fail, but if there's no report the key `CURRENTLY_VIEWED_REPORTID` would be set to `''`

marcaaron replied:
I don't see why this wouldn't work, but can also do some testing. I was a little confused on how it works at first, but after looking at the entire PR it makes sense. I have some thoughts on simplifying things.

kidroca replied:
I've left you a note on another thread

kidroca commented on src/libs/Navigation/NavigationRoot.js:
```diff
@@ -1,115 +1,52 @@
+            return;
+        }
 
-                    initialState.routes = [
-                        homeRoute,
-                        ...initialState.routes,
-                    ];
-                }
```
This was causing 2 home routes/states to be stored in the initial state, though I didn't see any negative effects resulting from this.

marcaaron replied:
Nice. I think it might have been left behind from an earlier iteration and agree it makes no sense.

kidroca replied:
And then I saw some negative effects 😃

kidroca commented on src/libs/Navigation/NavigationRoot.js:
```diff
@@ -1,115 +1,52 @@
     }
 
-    componentDidMount() {
-        Linking.getInitialURL()
-            .then((initialUrl) => {
-                // On web we should be able to parse this. It will be null on native for now until deep links are
-                // hooked up
-                const path = getPathName(initialUrl);
-                let initialState = getStateFromPath(path, linkingConfig.config);
-                setCurrentURL(path);
```
The entire logic here, seems no longer necessary - `setCurrentURL` is covered by the call in 
 `onStateChange`. 
- initial state is resolved correctly
- deep links are still resolved. E.g. try `localhost:8080/setpassword/10024706/EHOHMPGWM`

kidroca commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -124,7 +142,22 @@ class AuthScreens extends React.Component {
+    }
+
     render() {
+        if (!this.initialHomeParams) { return null; }
```
We might replace this with a proper loader here if the overall approach seem good. Doesn't seem necessary as even waiting for the fetch happens very fast

kidroca commented on src/libs/Navigation/NavigationRoot.js:
```diff
@@ -1,115 +1,52 @@
         return (
             <NavigationContainer
-                initialState={this.state.initialState}
-                onStateChange={(state) => {
-                    if (!state) {
-                        return;
-                    }
-
-                    const path = getPathFromState(state, linkingConfig.config);
-                    if (path.includes(ROUTES.REPORT)) {
-                        const reportID = Number(_.last(path.split('/')));
-                        if (reportID && !_.isNaN(reportID)) {
-                            updateCurrentlyViewedReportID(reportID);
-                        }
-                    }
-
-                    setCurrentURL(path);
-                }}
```
Moved this to a dedicated method

kidroca commented:
Found one more thing that this PR fixes. I don't think it was reported. It's Android only and related to the keyboard issues I've discovered and tracked to Navigation state:

#### Navigation returns to the same screen/modal on Android

https://user-images.githubusercontent.com/12156624/114439269-777bbc00-9bd1-11eb-8489-f9425422c096.mp4

I've stumbled on this while testing if my changes also cover: #1855 

It looks like this PR together with #2221 address fully the keyboard issues in #1855 and #2150 (and the above that's not yet reported) 🦸

marcaaron commented on src/libs/Navigation/NavigationRoot.js:
```diff
@@ -1,115 +1,52 @@
+     * Intercept state changes and perform different logic
+     * @param {NavigationState} state
+     */
+    handleStateChange(state) {
```
I like that this was moved, but our style guide says to not use `handle` prefix and instead pick a name that describes what the method does e.g. `updateStoredRoutes()` or whatever you like.

kidroca replied:
Yep, I know, just couldn't come up with a name.
I wasn't sure if this handler wouldn't grow in the future covering different state change cases so naming it after something now might not be relevant. E.g. the name I'm looking for is "something that decides what specific action should take place when a state changes" - it's just the decision maker and delegates to the specific action function

marcaaron replied:
Yeah it's just a style thing that we have for consistency. Since we are parsing the urls from the state we can make this `parseAndStoreRoute()` or something. If the responsibility grows we can change the name.

marcaaron commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -69,6 +74,10 @@ class AuthScreens extends React.Component {
+
+        if (props.initialReportID) {
+            this.setInitialHomeParams(props.initialReportID);
+        }
```
I think we could maybe simplify this a bit by just setting `this.initialReportID` instead of a method to do this for us?

marcaaron commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -109,11 +118,20 @@ class AuthScreens extends React.Component {
+        // Update initialHomeParams only once
+        if (!this.initialHomeParams) {
+            // Either we have a reportID or fetchAllReports resolved with no reports. Otherwise keep waiting
+            if (nextProps.initialReportID || nextProps.initialReportID === '') {
```
If someone e.g. decided to give a `defaultProps` of empty string to `initialReportID` this would break right?

I don't have a better suggestion yet, but I'm curious if we can make more obvious what is happening here as `initialReportID === ''` feels like an implementation detail for `fetchAllReports resolved with no reports`.

kidroca replied:
Yeah, I thought about this too, but decided to get some feedback from you and review my general direction on this PR

We could hook to the promise here https://github.com/Expensify/Expensify.cash/blob/0874880f20c22f4453a4759e6885e7f28f170ca2/src/libs/Navigation/AppNavigator/AuthScreens.js#L106

And do something like

```js
componentDidMount() {
  fetchAllReports(true, true).finally(() => this.resolveInitialState);
}

resolveInitialState() {
  // If initial nav params haven't resolved up until now just start with sensible defaults
  const routeID = this.props.initialReportID || '';

  // we can return early if initial state is already set
  
  // We'll need to use `setState` to ensure a re-render
  this.setState({
     initialHomeParams: {
        screen: 'Report',
        params: {reportID},
     }
  });
}
```

This above is definitely more explicit than interpreting `nextProps.initialReportID === ''` and comments

shouldComponentUpdate - should account for state updates using the 2nd parameter

marcaaron replied:
Hmm yeah this is definitely more explicit and probably how I'd approach this as well, but we also await the promise now (which we have set limits on). I'm unsure if we'd want to break the streak for this.

Another option would be to create something like a `isLoadingInitialReports` flag in Onyx,  but this seems kind of backwards so maybe we can just live with this for now.

kidroca replied:
Sorry but can you please say what do you mean by "this" so I can address it tomorrow
- use the "promise" version 
- leave it the way it is for the time being?

marcaaron replied:
No worries, I am saying to leave it like we have it here. We can always add more comments to fill out additional context that might not be obvious.

marcaaron reviewed (commented):
Just leaving some initial thoughts since this is in draft.

kidroca committed:
refactor: AuthScreens address PR comments

kidroca committed:
refactor: NavigationRoot address PR comments

kidroca committed:
refactor: NavigationRoot move viewed report persistence logic

The Report Screen renders the "currently viewed report" - it can persist it
This way we doesn't have to check per each navigation state change

kidroca commented on src/pages/home/ReportScreen.js:
```diff
@@ -37,6 +38,7 @@ class ReportScreen extends React.Component {
 
         if (reportChanged) {
             this.prepareTransition();
+            this.storeCurrentlyViewedReport();
```
This replaces the logic here:

https://github.com/Expensify/Expensify.cash/blob/d3f1d28478357e2a66b155e660def191d6ed79a8/src/libs/Navigation/NavigationRoot.js#L104-L109

The `ReportScreen` renders the last viewed report so we don't need to check per each navigation change, but only save the last report that was viewed in `ReportScreen`

marcaaron replied:
This change makes sense to me. I don't see how it's different from what we had before (maybe a bit cleaner). Was there any other motivation for the change or just cleaning stuff up?

kidroca replied:
Primarily clean up, it should still work the same as before. I can revert this you think it might cause some issue down the road

kidroca commented on src/pages/home/ReportScreen.js:
```diff
@@ -73,6 +75,11 @@ class ReportScreen extends React.Component {
     }
 
+    storeCurrentlyViewedReport() {
+        const reportID = this.getReportID();
+        updateCurrentlyViewedReportID(reportID);
```
We don't need an `if` guard here as currently it's guaranteed that `reportID` would be a proper value

This function would not run for the initial render but only for updates, which guarantees `reportID` would exist. For the initial render we either had something in Onyx or never had anything anyway.

Anyway I can add a safety check if you'd like

kidroca committed:
Merge remote-tracking branch 'e.cash/master' into kidroca/fix-url-not-updating-after-login-or-refresh

Conflicts:
	src/libs/Navigation/AppNavigator/AuthScreens.js

kidroca commented:
Updated. 
Merged latest changes from master and resolved conflicts

marcaaron commented on src/pages/home/ReportScreen.js:
```diff
@@ -73,6 +75,11 @@ class ReportScreen extends React.Component {
         this.loadingTimerId = setTimeout(() => this.setState({isLoading: false}), 300);
     }
 
+    storeCurrentlyViewedReport() {
```
All our components methods must have a method doc even if they don't take params or return anything.

kidroca replied:
> Avoid docs that don't add any additional information.
> Method descriptions are optional and should be added when it's not obvious what the purpose of the method is.
https://github.com/Expensify/Expensify.cash/blob/master/STYLE.md#jsdocs 🤓 

This is a class method and I'll just be adding this as a description: "Persists the currently viewed report id". I could extend it to "...with Onyx" but this is just an implementation detail 

Anyway this is not the first time, someone would bring up func/method documentation. In fact it is always the case that I should add documentation so you might consider updating the guide - I would have just always added documentation if I hadn't read the above in the style guide.

marcaaron replied:
Thanks for calling that out. I will update it to be more clear.

marcaaron reviewed (changes requested):
Changes look great to me just one small comment left about adding a method doc.

kidroca committed:
docs: ReportScreen add method documentation

kidroca commented:
Updated

marcaaron reviewed (approved):
LGTM! 👍 

I'm not sure if we need/want to do anything about this yet. But an interesting side effect of this change is that on a fresh sign in the report with Concierge loads first on web, but the header component does not say we are chatting with Concierge.

![2021-04-13_09-37-54](https://user-images.githubusercontent.com/32969087/114611390-b2ffa000-9c3c-11eb-8781-eb49f4801b8b.png)

My guess would be this happens because the `HeaderView` is still waiting for the `personalDetails` to be set. And ideally we would wait for both of these things before showing the very first report.

Compared to the main branch where we wait (but also never redirect which is what we are fixing)
![2021-04-13_09-43-56](https://user-images.githubusercontent.com/32969087/114611535-db879a00-9c3c-11eb-9270-ed5380ea3952.png)

I'd propose making a follow up issue for this.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

kidroca commented:
Strange, on my sample video for web above the header information loads/updates as well. Though I made a new user and there's only the Concierge chat. I'll have a look tomorrow

marcaaron commented:
Perhaps because there are very few users and everything loads fast.

kidroca commented:
@marcaaron 
> I'm not sure if we need/want to do anything about this yet. But an **interesting side effect of this change** is that on a fresh sign in the report with Concierge loads first on web, but the header component does not say we are chatting with Concierge.

I thought you mean the header never updated and stayed the way it is in the your screenshot 
I'm just saying that it eventually loads

**It's not a side effect of this PR** or the loader addition - it was always the case that elements would render in their own time as soon as they have data:

#### A video from 1359f9fc before any of the loader/url fixes

https://user-images.githubusercontent.com/12156624/114670423-a39b5800-9d0b-11eb-9c6c-ee0ffd88e8ed.mp4
- the header render blank then loads

#### A video from current state d10d52ca

https://user-images.githubusercontent.com/12156624/114670527-c3328080-9d0b-11eb-82fd-f4fb3a838c52.mp4

To fine tune loading the appearance of the loader we could replace the timeout logic with state change polling like it's done here:  kidroca@ae376b4

marcaaron commented:
>it was always the case that elements would render in their own time as soon as they have data

Apologies, sounds like the case here. To clarify, I didn't mean to suggest it should be fixed as a follow up or anything (but looks like it re-reading my comment 🤦). Perhaps more accurate to say the effect or more noticeable/pronounced now.

isagoico commented:
@marcaaron Question about QAing this. I think we're only able to execute the steps in Web and mWeb to be able to check the URL. Is there a way to test this on Apps (android, iOS and desktop) or should we go ahead with just web/mweb?

marcaaron commented:
Thanks @isagoico, it should be testable by using deep links. We're using universal links so in most cases if we can reach a link like `https://expensify.cash/settings` then as long as the mobile app is installed when opened from a device it should deep link to the correct place in the app.

marcaaron commented:
Of course, there is no "url" to verify but we can see if things are working at least.

isagoico commented:
@marcaaron Links are not working for the iOS app. It only open mWeb and doesn't show the option to open in cash app 🤔 Android is working as expected. Should I go ahead an open a separate issue for this?

isagoico commented:
The `Create Account` scenario is failing because of this issue: 

## Sign Up - Blank screen after setting the password 

## Expected result
User proceeds to expensify.cash chat list

## Actual Result
User is redirected to a blank screen and a JS error can be seen in the console

## Action Performed
1. Open staging E.cash URL
2. While logged out, type and login with brand new email address so it can trigger sign up flow
3. Get validation link from e-mail, change it to staging accordingly
4. Type new password and click "set password"

## Platform
Issue is confirmed in: 

Android ✔️ 
Web ✔️ 

**Build:** 1.0.21-0
**Notes/images/video:**
Issue is not reproducible in production

[Video](https://utest-dl.s3.amazonaws.com/12102/26469/311617/null/bugAttachment/Bug5020701_video.mp4?AWSAccessKeyId=AKIAJ2UIWMJ2OMC3UCQQ&Expires=1649973302&Signature=BQrcmfY1W2b7OGSKFCZd%2BnKAenU%3D
)
[Console log](https://utest-dl.s3.amazonaws.com/12102/26469/311617/null/bugAttachment/Bug5020701_staging.expensify.cash-1618436983536.log?AWSAccessKeyId=AKIAJ2UIWMJ2OMC3UCQQ&Expires=1649973302&Signature=%2BoTAdvwSec5jIt0re83gAPKBn3o%3D)

![image](https://user-images.githubusercontent.com/44479856/114785684-0a8c3180-9d4b-11eb-89e2-0784d7ba383f.png)

isagoico commented:
I'm not 100% sure if this PR is the one responsible for this behaviour so please let me know if it's not the case so I can open a separate issue :)

marcaaron commented:
Not sure, but I'm not able to reproduce that issue locally.

kidroca commented:
> I'm not 100% sure if this PR is the one responsible for this behaviour so please let me know if it's not the case so I can open a separate issue :)

I think it kinda is 😭 

@marcaaron 
It probably doesn't happen all the time but in the shared video you can see the line of code in Report.js where the exception was triggered: [Report.js:601](https://github.com/Expensify/Expensify.cash/blob/fda2bab095281de6383b4aaa6c9f07c433d0835e/src/libs/actions/Report.js#L601)

I've traced this to 

1. `AuthScreens` mounts
2. Invokes `fetchAllReports` -> Reports.fetchAll
3. Reports.fetchAll -> actually calls Reports.fetchChatReports
4. There's a case that this leads to Reports.fetchOrCreateChatReport
5. Which will execute the code on line 601

This would happen milliseconds before `fetchAll` sets `CURRENTLY_VIEWED_REPORTID` and mount the RootStack

It doesn't have to perform navigation for this specific case, as it's being resolved from initial params, but I see the function is used on a couple of other places, to create a new chat and redirect navigation to it.

For the moment it can be addressed as suggested here: https://reactnavigation.org/docs/navigating-without-navigation-prop/#handling-initialization
> If you try to navigate without rendering a navigator or before the navigator finishes mounting, it will throw and crash your app if not handled. So you'll need to add an additional check to decide what to do until your app mounts.

We can update `Navigation.navigate` so that it doesn't try to perform navigation while no Router is mounted. This will address the case and other errors that might result from this brief period when no navigator is mounted

An alternative solution might be to only switch to `AuthScreens` when initial data is ready, so the swap between navigators can be instant, but this means some of the fetching would need to be moved out of there. 
And it still doesn't guarantee if navigation event was triggered from the Public stack but in the meantime we're switching to the Auth stack what would happen...

Or maybe simply calling `Onyx.set(ONYXKEYS.CURRENTLY_VIEWED_REPORTID, reportId);` before the `Navigation.navigate` on 601 would be enough

marcaaron commented:
Ah interesting. I see what you're saying here. So we need to do something like:

1. Allow `NavigationContainer` to initialize with no report
2. Record when the `NavigationContainer` is ready
3. Allow navigation attempts once things are "ready"

I'm not sure what would be best in this case, but one possibly naive idea would be to:

add on `onReady={onNavigatorReady}` to the `NavigationContainer` and the callback would be exported from `Navigation`. If any attempts to navigate are made before called we can put the route in a temp variable and then call them once the navigator is ready.

```js
import {onNavigationContainerReady} from '/Navigation'

<NavigationContainer
    onReady={onNavigationContainerReady}
```

```js
let lastAttemptedRoute;
let isReady = false;

function navigate(route) {
    if (!isReady) {
        lastAttemptedRoute = route;
        return
    }
    ...
}

function onNavigationContainerReady() {
    isReady = true;
    navigate(lastAttemptedRoute);
}
```

kidroca commented:
In their documentation they would just skip doing anything. Your solution seem an improvement over that. Looks like it would work 

It might be a bit too clever but I you can do something like this:

```js
let lastAttemptedRoute;

function navigate(route) {
 // unmodified
}

function navigateLater(route) {
  lastAttemptedRoute = route;
}

function onNavigationContainerReady() {
    navigate(lastAttemptedRoute);
    exports.navigate = navigate;
}

const exports = {
  navigate: navigateLater,
  onNavigationContainerReady,
};

export default exports;
```

Because that check is only needed initially but will be performed for the lifetime of the Navigation
Also it gets in the way of reading the actual code

marcaaron commented:
Yep I like that improvement.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash
    /actions/runs/765130740) 🚀 to 
    staging on Mon Apr 19 2021 at 22:41:56 GMT+0000 (Coordinated Universal Time)

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

kidroca commented:
Hey @marcaaron I'm not sure if you expect anything more from me? Should I open a PR add work on a fix?

marcaaron commented:
If you want to create an issue for this follow up we can make a new job out of it ?

kidroca commented:
That would be great! 
I can use the requirements you listed above and make a ticket: https://github.com/Expensify/Expensify.cash/pull/2346#issuecomment-820604605

marcaaron commented:
Sounds good thanks :)

roryabraham commented:
Hi guys! Just noticing that this pull request is still marked as a deploy blocker.

@isagoico, can you please confirm if the issue describe [here](https://github.com/Expensify/Expensify.cash/pull/2346#issuecomment-819875664) is still present on staging but not production? If so, I think we should create a new issue and get it resolved ASAP. Thanks!

kidroca commented:
The new issue is already created, the link is just above your comment

roryabraham commented:
Thanks @kidroca, I am going to remove the `DeployBlockerCash` label from this pull request and get it added to that other issue.
